File: example/src/main/java/com/github/vivchar/example/pages/simple/LoadMoreFragment.java
package com.github.vivchar.example.pages.simple;

import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.github.vivchar.example.BaseScreenFragment;
import com.github.vivchar.example.R;
import com.github.vivchar.example.widgets.EndlessScrollListener;
import com.github.vivchar.example.widgets.ItemOffsetDecoration;
import com.github.vivchar.rendererrecyclerviewadapter.binder.LoadMoreViewBinder;
import com.github.vivchar.rendererrecyclerviewadapter.RendererRecyclerViewAdapter;
import com.github.vivchar.rendererrecyclerviewadapter.ViewModel;
import com.github.vivchar.rendererrecyclerviewadapter.binder.ViewBinder;

import java.lang.reflect.Type;

/**
 * Created by Vivchar Vitaly on 12/29/17.
 */

public class LoadMoreFragment extends BaseScreenFragment {

	public static final int COLUMNS_COUNT = 4;
	private final YourDataProvider mYourDataProvider = new YourDataProvider();
	private RendererRecyclerViewAdapter mAdapter;
	private RecyclerView mRecyclerView;
	private GridLayoutManager mLayoutManager;

	@Nullable
	@Override
	public View onCreateView(final LayoutInflater inflater,
	                         @Nullable final ViewGroup container,
	                         @Nullable final Bundle savedInstanceState) {

		final View view = inflater.inflate(R.layout.fragment_list, container, false);

		mAdapter = new RendererRecyclerViewAdapter();
		mAdapter.enableDiffUtil();


//		mAdapter.setLoadMoreModel(new YourLoadMoreModel()); /* you can change the LoadMoreModel if needed */
		mAdapter.registerRenderer(new LoadMoreViewBinder(R.layout.item_load_more));
		mAdapter.registerRenderer(new ViewBinder<>(R.layout.item_simple_square, SimpleViewModel.class,
				(model, finder, payloads) -> finder.setText(R.id.text, model.getText())
		));
//		adapter.registerRenderer(...);
//		adapter.registerRenderer(...);

		mAdapter.setItems(mYourDataProvider.getLoadMoreItems());

		mLayoutManager = new GridLayoutManager(getContext(), COLUMNS_COUNT);
		mLayoutManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
			@Override
			public int getSpanSize(final int position) {
				final Type type = mAdapter.getType(position);
				if (type.equals(SimpleViewModel.class)) {
					return 1;
				}
				return COLUMNS_COUNT;
			}
		});

		mRecyclerView = (RecyclerView) view.findViewById(R.id.recycler_view);
		mRecyclerView.setAdapter(mAdapter);
		mRecyclerView.setLayoutManager(mLayoutManager);
		mRecyclerView.addItemDecoration(new ItemOffsetDecoration(10));
		mRecyclerView.addOnScrollListener(new EndlessScrollListener() {
			@Override
			public void onLoadMore(final int page, final int totalItemsCount) {
				Log.d("#####", "onLoadMore " + page);
				mAdapter.showLoadMore();
//				mAdapter.hideLoadMore(); /* if you need force hide progress or call setItems() */
				mYourDataProvider.getLoadMoreItems(list -> getActivity().runOnUiThread(() -> mAdapter.setItems(list)));
			}
		});

		return view;
	}

	public static class SimpleViewModel implements ViewModel {

		private final String mText;

		public SimpleViewModel(final String text) {
			mText = text;
		}

		public String getText() {
			return mText;
		}
	}
}

File: example/src/main/java/com/github/vivchar/example/pages/simple/LoadMoreFragment.kt
package com.github.vivchar.example.pages.simple

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.GridLayoutManager
import com.github.vivchar.example.R
import com.github.vivchar.example.base.BaseFragment
import com.github.vivchar.example.databinding.FragmentListBinding
import com.github.vivchar.example.widgets.EndlessScrollListener
import com.github.vivchar.example.widgets.ItemOffsetDecoration
import com.github.vivchar.example.widgets.MyAdapter
import com.github.vivchar.rendererrecyclerviewadapter.LoadMoreViewBinder
import com.github.vivchar.rendererrecyclerviewadapter.ViewFinder
import com.github.vivchar.rendererrecyclerviewadapter.ViewModel
import com.github.vivchar.rendererrecyclerviewadapter.ViewRenderer
import kotlinx.coroutines.launch

class LoadMoreFragment : BaseFragment<FragmentListBinding>() {

	private val viewModel: LoadMoreViewModelVM by viewModels()
	private val adapter = MyAdapter()

	override fun createBinding(inflater: LayoutInflater, container: ViewGroup?) =
		FragmentListBinding.inflate(inflater, container, false)

	override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
		super.onViewCreated(view, savedInstanceState)

		adapter.enableDiffUtil()
		adapter.registerRenderer(LoadMoreViewBinder(R.layout.item_load_more))
		adapter.registerRenderer(
			ViewRenderer<SimpleViewModel, ViewFinder>(
				R.layout.item_simple_square, SimpleViewModel::class.java
			) { model, finder, _ -> finder.setText(R.id.text, model.text) }
		)

		val layoutManager = GridLayoutManager(context, COLUMNS_COUNT)
		layoutManager.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
			override fun getSpanSize(position: Int): Int {
				return when (adapter.getType(position)) {
					SimpleViewModel::class.java -> 1
					else -> COLUMNS_COUNT
				}
			}
		}

		binding.recyclerView.adapter = adapter
		binding.recyclerView.layoutManager = layoutManager
		binding.recyclerView.addItemDecoration(ItemOffsetDecoration(10))
		binding.recyclerView.addOnScrollListener(object : EndlessScrollListener() {
			override fun onLoadMore(page: Int, totalItemsCount: Int) {
				viewModel.onLoadMore()
			}
		})

		viewLifecycleOwner.lifecycleScope.launch {
			viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
				viewModel.state.collect { state ->
					if (state.showLoadMore) {
						adapter.showLoadMore()
					} else {
						adapter.setItems(state.items)
					}
				}
			}
		}
	}

	data class SimpleViewModel(val text: String) : ViewModel

	companion object {
		const val COLUMNS_COUNT = 4
	}
}
